### PR TITLE
Inherited methods

### PR DESCRIPTION
**Pull request by dbrosoft**
Generate inherited methods on which the index file indicates that annotations are to be inserted.  
A bug involving annotations on inner types remains to be fixed.  (I will push an expanded test set shortly.)  I would suggest turning off this functionality by default pending further testing; I can add a flag to turn it back on.

### CODE REVIEW

File: annotation-file-utilities/src/annotator/find/ASTPathCriterion.java [hunk 1]
```diff
@@ -96,9 +95,7 @@ public boolean isSatisfiedBy(TreePath path, Tree leaf) {
     /** {@inheritDoc} */
     @Override
     public boolean isSatisfiedBy(TreePath path) {
-        if (path == null) {
-            return false;
-        }
+        if (path == null || astPath.isEmpty()) { return false; }
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/ASTPathCriterion.java]: Please don't put actions on same line as conditions.  It's harder to read, it is not the standard Java style, and it is inconsistent with elsewhere in the file.  Furthermore, it makes the diffs larger.

Comment (mernst) [inline annotation-file-utilities/src/annotator/find/ASTPathCriterion.java]: This comment still has not been addressed.

Comment (dbrosoft) [inline annotation-file-utilities/src/annotator/find/ASTPathCriterion.java]: Fixed.

**Comment by wmdietl** [comment]
Hi Dan, are you looking for comments on this?

**Comment by dbrosoft** [comment]
Sure, I would appreciate your feedback.

-- Dan.

File: annotation-file-utilities/src/annotator/find/ASTPathCriterion.java [hunk 2]
```diff
@@ -140,7 +135,7 @@ public boolean isSatisfiedBy(TreePath path) {
 
         int astPathLen = astPath.size();
         int actualPathLen = actualPath.size();
-        if (astPathLen == 0 || actualPathLen == 0) { return false; }
+        if (actualPathLen <= 0) { return false; }
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/ASTPathCriterion.java]: How can actualPathLen, which is actualPath.size() where actualPath is a List, be negative?  This looks like a wrong test to me.

File: annotation-file-utilities/src/annotator/find/InMethodCriterion.java [hunk 3]
```diff
@@ -55,9 +67,9 @@ public boolean isSatisfiedBy(TreePath path) {
         staticDecl = mods.getFlags().contains(Modifier.STATIC);
       }
       path = path.getParentPath();
-    } while (path != null && path.getLeaf() != null);
+    } while (path != null);
 
-    result = (staticDecl ? "<clinit>()V" : "<init>()V").equals(name);
+    result |= (staticDecl ? "<clinit>()V" : "<init>()V").equals(name);
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/InMethodCriterion.java]: | is a bitwise operator.  Use || instead, as it is a boolean operator.

File: annotation-file-utilities/src/annotator/find/Criteria.java [hunk 4]
```diff
@@ -96,6 +96,14 @@ public boolean isSatisfiedBy(TreePath path) {
    * @return true iff this is the criteria on a receiver
    */
   public boolean isOnReceiver() {
+    ASTPath astPath = getASTPath();
+    if (astPath != null && !astPath.isEmpty()) {
+      ASTPath.ASTEntry entry = astPath.get(-1);
+      return entry.getTreeKind() == Tree.Kind.METHOD
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/Criteria.java]: Is it possible that there is no ASTPathCriterion in this, and yet this contains some criterion whose kind is RECEIVER?  If so, this return statement is incorrect.  The code should return true if the condition is satisfied, and otherwise fall through to the test immediately below (which is currently on line 107).  The same comment applies to other code in this class, which I think should also be changed.

File: annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.LinkedHashMultimap;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+
+import com.sun.tools.javac.code.Symbol;
+import com.sun.tools.javac.model.JavacElements;
+
+/**
+ * Provides static access to JVML descriptors of methods for which
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java]: What is "static access"?  It would be clearer to say that this class doesn't represent anything and is just a collection of static methods.  Furthermore, add a private constructor that throws an exception, so clients can't accidentally instantiate it.

File: annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+ *  implementations are inherited in a given class.
+ *
+ * @author dbro
+ */
+public class InheritedSymbolFinder {
+  private static Source source = null;
+
+  // class (flat) name -> method index
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java]: Make these Javadoc comments.  (There is a Javadoc mode that also documents private members, and IDEs can also use the information.)

Also, what is a "flat name"?  That's not standard terminology and I feel it would be better to use terms with well-known meanings.

Finally, this comment has the wrong type.  The map is String -> String -> Element, but your documentation doesn't have the right number of ->.

File: annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+  // keys: class (flat) name
+  // values: multimap of method (simple) name to matching method elems
+  private static Map<String, Multimap<String, ExecutableElement>> cash =
+      new HashMap<String, Multimap<String, ExecutableElement>>();
+
+  // initialized once and for all
+  private static Multimap<String, ExecutableElement> objectMethods = null;
+
+  private InheritedSymbolFinder() {}
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java]: This should throw an exception rather than do nothing.  I would also put it at the top of the class.  (For ordinary classes, putting the constructor after the fields is reasonable.)

File: annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+  // class (flat) name -> method index
+  private static Map<String, Multimap<String, ExecutableElement>> cache =
+      new TreeMap<String, Multimap<String, ExecutableElement>>();
+
+  // keys: class (flat) name
+  // values: multimap of method (simple) name to matching method elems
+  private static Map<String, Multimap<String, ExecutableElement>> cash =
+      new HashMap<String, Multimap<String, ExecutableElement>>();
+
+  // initialized once and for all
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java]: I don't understand this comment.  Especially since it is initialized to null -- is it never reset?  If so, why is this field even needed?

File: annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+ * Provides static access to JVML descriptors of methods for which
+ *  implementations are inherited in a given class.
+ *
+ * @author dbro
+ */
+public class InheritedSymbolFinder {
+  private static Source source = null;
+
+  // class (flat) name -> method index
+  private static Map<String, Multimap<String, ExecutableElement>> cache =
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java]: "cache" is a weak name.  Choose a name that reflects what this means.

File: annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+      new TreeMap<String, Multimap<String, ExecutableElement>>();
+
+  // keys: class (flat) name
+  // values: multimap of method (simple) name to matching method elems
+  private static Map<String, Multimap<String, ExecutableElement>> cash =
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java]: What does this name mean?  If it's a corruption of "cache", that's bad:  code shouldn't have subtle, hard-to-understand puns but should be clearly named.

File: annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+      new HashMap<String, Multimap<String, ExecutableElement>>();
+
+  // initialized once and for all
+  private static Multimap<String, ExecutableElement> objectMethods = null;
+
+  private InheritedSymbolFinder() {}
+
+  // Needed for obtaining instances of utility classes Types and Elements.
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java]: Every method should have a comment that says what it does.  This one talks about why clients might call it (though not clearly), but it doesn't say what the method does.

File: annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+  // TODO: make methods non-static and pass source to constructor instead.
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java]: Do you plan to address this TODO?  Doing so seems straightforward and valuable.

File: annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+          }
+        }
+        cash.put("java.lang.Object", objectMethods);
+      }
+    }
+  }
+
+  /**
+   * Finds whether a method matching a JVML descriptor has an inherited
+   *  implementation in a given class.
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java]: Don't put extra indentation on the second line.  You did this in some other places, too, so please fix them all.

File: annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+   * @return whether a matching method exists in the class
+   */
+  public static boolean isInheritedIn(Symbol.ClassSymbol classSymbol,
+      String methodDescriptor) {
+    Multimap<String, ExecutableElement> inherited = getInherited(classSymbol);
+    int ix = methodDescriptor.indexOf('(');
+    String key = methodDescriptor.substring(0, ix);
+    Collection<ExecutableElement> elems = inherited.get(key);
+
+    // There may be multiple methods stored under the same key (bare
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java]: Please use standard terminology rather than making up new terms.  By "bare method name" do you mean "simple method name"?

File: annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+    // method name), so check each of them.
+    // TODO: figure out good way to determine whether method descriptor
+    // gives the type of an inherited method.  (Only exact matches work
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java]: By "exact matches", do you mean that the formal parameters must match exactly, or that the formal parameters and return type must match exactly, or something else?  You only need to be checking the formal parameters, not the return type.

File: annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+    // now.)
+    if (elems != null) {
+      for (ExecutableElement elem : elems) {
+        String jvmName = JVMNames.getJVMMethodName(elem);
+        int ix0 = methodDescriptor.lastIndexOf(')') + 1;
+        int ix1 = jvmName.lastIndexOf(')') + 1;
+        String key0 = methodDescriptor.substring(0, ix0);
+        String key1 = jvmName.substring(0, ix1);
+        if (key0.equals(key1)) { return true; }
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java]: Although it takes a bit more vertical space, it's standard Java style to put the then body on a separate line even if it's only one line long.  Please do so, here and elsewhere.

File: annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+    }
+    return false;
+  }
+
+  /**
+   * Finds inherited methods defined for a class (given by its symbol
+   *  table entry).
+   *
+   * @param classSymbol symbol table entry for a class
+   * @return JVML descriptors of methods in scope of class
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java]: This `@return` comment seems wrong.  It sounds like it is describing a list, but the return value is a map.  Can you better describe the return value?

File: annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+   */
+  private static Multimap<String, ExecutableElement>
+  getInherited(final Symbol.ClassSymbol classSymbol) {
+    if (source == null) { throw new IllegalStateException("null source"); }
+
+    final JavacElements elements = (JavacElements) source.getElementUtils();
+    final Types types = source.getTypeUtils();
+    String className = classSymbol.flatName().toString();
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java]: Is there a reason that the previous two variables were final but this one is not?  The contrast makes a reader think there is a distinction being made.

File: annotation-file-utilities/src/annotator/find/Insertion.java [hunk 19]
```diff
@@ -50,22 +50,66 @@
...
      */
     protected static Set<String> alwaysQualify = new LinkedHashSet<String>();
 
     /**
+     * Type to be inserted, if applicable; otherwise null.
+     */
+    protected Type type;
+
+    /**
+     * If true only the annotations from {@link type} will be inserted.
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/Insertion.java]: What is the alternative?  What other things get inserted if this is false?

File: annotation-file-utilities/src/annotator/find/Insertion.java [hunk 20]
```diff
@@ -50,22 +50,66 @@
...
+     */
+    public Insertion(Type type, Criteria criteria,
+            List<Insertion> innerTypeInsertions) {
+        this(type, criteria, false, innerTypeInsertions);
+    }
+
+    /**
+     * Creates a new insertion.
+     *
+     * @param type type to be generated, if applicable
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/Insertion.java]: What does it mean to "generate" a type?  How is that related to insertion?

File: annotation-file-utilities/src/annotator/find/Insertion.java [hunk 21]
```diff
@@ -50,22 +50,66 @@
     protected Set<String> packageNames;
 
     /**
+     * The inner types to go on this insertion. See {@link ReceiverInsertion}
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/Insertion.java]: Are the insertions types (as the comment says) or possibly other things such as annotations and receiver declarations?

File: annotation-file-utilities/src/annotator/find/Insertion.java [hunk 22]
```diff
@@ -203,6 +247,50 @@ public static void setAlwaysQualify(Set<String> set) {
     }
 
     /**
+     * Gets the type.  It is assumed that the returned value will be
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/Insertion.java]: Avoid passive voice.  Do you mean, "The client must modify the returned value to update the type to be inserted."?
How does the client do this, and why?  Is it always required or only sometimes?

Also, rather than just "the type", I think it would be clearer to say "the type to be inserted, if any".  Similarly clarify `setType` below.

File: annotation-file-utilities/src/annotator/find/Insertion.java [hunk 23]
```diff
@@ -203,6 +247,50 @@ public static void setAlwaysQualify(Set<String> set) {
...
+    }
+
+    /**
+     * Sets the type.
+     */
+    public void setType(Type type) {
+        this.type = type;
+    }
+
+    public DeclaredType getBaseType() {
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/Insertion.java]: Every method needs Javadoc documentation.  Please write it for this method and any other methods that lack it.  I see some more in this class, and others might exist elsewhere in the pull request.

File: annotation-file-utilities/src/annotator/find/Insertion.java [hunk 24]
```diff
@@ -203,6 +247,50 @@ public static void setAlwaysQualify(Set<String> set) {
...
+            case ARRAY:
+                return getBaseType(((ArrayType) type).getComponentType());
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Indicates whether inserted text should consist <em>only</em> of
+     *  the indicated annotations.  Useful when subordinate insertions
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/Insertion.java]: Can you clarify "subordinate insertions", or give an example or two of a subordinate insertion?

File: annotation-file-utilities/src/annotator/find/Insertion.java [hunk 25]
```diff
@@ -389,6 +485,20 @@ public static void decorateType(List<Insertion> innerTypeInsertions, final Type
...
+     * Adds each of the given inner type insertions to the correct part of the
+     * type, based on the insertion's type path.
+     *
+     * @param innerTypeInsertions
+     *          The insertions to add to the type. These must be inner type
+     *          insertions, meaning each of the insertions' {@link Criteria}
+     *          must contain a {@link GenericArrayLocationCriterion} and
+     *          {@link GenericArrayLocationCriterion#getLocation()} must return
+     *          a non-empty list.
+     * @param outerType The type to add the insertions to.
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/Insertion.java]: `@param` text should not end with punctuation (or start with a capitalized letter), if it is a single sentence fragment.

File: annotation-file-utilities/src/annotator/find/Insertion.java [hunk 26]
```diff
@@ -481,9 +591,9 @@ public static void decorateType(List<Insertion> innerTypeInsertions,
         }
     }
 
+    // auxiliary for previous method
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/Insertion.java]: What does it do?  How is it different?

File: annotation-file-utilities/src/annotator/find/NewInsertion.java [hunk 27]
```diff
@@ -11,7 +11,7 @@
  * @author dbro
  *
  */
-public class NewInsertion extends TypedInsertion {
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/NewInsertion.java]: Please add Javadoc for this class.

File: annotation-file-utilities/src/annotator/find/IsSigMethodCriterion.java [hunk 28]
```diff
@@ -403,6 +382,16 @@ public boolean isSatisfiedBy(TreePath path) {
     return true;
   }
 
+  /** Find a type parameter's bound in the current context. */
+  private String erasure(TypeParameterTree param) {
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/IsSigMethodCriterion.java]: "Erasure" also means dropping type parameters as in Map<String, Integer> gets erased to Map.  Would a name that relates to "bound" be better than "erasure"?

File: annotation-file-utilities/src/annotator/find/ParamCriterion.java [hunk 29]
```diff
@@ -10,8 +10,8 @@
 
 public class ParamCriterion implements Criterion {
 
-  private final String methodName;
-  private final Integer paramPos;
+  final String methodName;
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/ParamCriterion.java]: Please write brief documentation for these fields.  (For example, is the integer index zero-based or one-based?

File: annotation-file-utilities/src/annotator/find/ParamCriterion.java [hunk 30]
```diff
@@ -10,8 +10,8 @@
...
+  final Integer paramPos;
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/ParamCriterion.java]: Why is this an Integer rather than an int?

File: annotation-file-utilities/src/annotator/find/ReturnTypeCriterion.java [hunk 31]
```diff
@@ -34,6 +34,10 @@ public boolean isSatisfiedBy(TreePath path) {
     Criteria.dbug.debug("ReturnTypeCriterion.isSatisfiedBy(%s); this=%n",
         Main.pathToString(path), this.toString());
 
+    if (CommonScanner.hasClassKind(path.getLeaf())) { return false; }
+    //JCTree.JCClassDecl decl = (JCTree.JCClassDecl) path.getLeaf();
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/ReturnTypeCriterion.java]: Is this commented-out code necessary?

File: annotation-file-utilities/src/annotator/find/TreeFinder.java [hunk 32]
```diff
@@ -253,50 +253,77 @@ private ASTRecord astRecord(Tree node) {
...
+      return Pair.of(astRecord(t), t.getStartPosition());
     }
 
     private Pair<ASTRecord, Integer> pathAndPos(JCTree t, int i) {
       return Pair.of(astRecord(t), i);
     }
 
-    /** @param t an expression for a type */
-    private Pair<ASTRecord, Integer> getBaseTypePosition(JCTree t) {
+    /** @param node an expression for a type */
```
Comment (mernst) [inline annotation-file-utilities/src/annotator/find/TreeFinder.java]: What is the return value?

File: annotation-file-utilities/src/annotator/find/ASTPathCriterion.java (hunk 2 above)
Comment (dbrosoft) [inline annotation-file-utilities/src/annotator/find/ASTPathCriterion.java]: True, in the current code `actualPathLen` cannot be negative, so `==` would work.  However,`<=` more precisely expresses the required invariant -- that only a positive integer is valid here -- independently from the size() invariant, so I find it preferable.  (If you strongly disagree, I will change it.)

File: annotation-file-utilities/src/annotator/find/Criteria.java (hunk 4 above)
Comment (dbrosoft) [inline annotation-file-utilities/src/annotator/find/Criteria.java]: If there is no ASTPathCriterion, then the result of `getASTPath()` is null, and control passes to the test below.

File: annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java (hunk 8 above)
Comment (dbrosoft) [inline annotation-file-utilities/src/annotator/find/InheritedSymbolFinder.java]: Will do, and in doing so will address the other issues with this source file.  (When I created this class, I was unaware that I would need a `Types` instance, which I could acquire only by using information then only present in `Source`.)

File: annotation-file-utilities/src/annotator/find/IsSigMethodCriterion.java (hunk 28 above)
Comment (dbrosoft) [inline annotation-file-utilities/src/annotator/find/IsSigMethodCriterion.java]: Yes.  I will rename the method "erasedBound".

File: annotation-file-utilities/src/annotator/find/ParamCriterion.java (hunk 30 above)
Comment (dbrosoft) [inline annotation-file-utilities/src/annotator/find/ParamCriterion.java]: I don't know.  I suspect it was like that before I began working on AFU.

File: annotation-file-utilities/src/annotator/find/ReturnTypeCriterion.java (hunk 31 above)
Comment (dbrosoft) [inline annotation-file-utilities/src/annotator/find/ReturnTypeCriterion.java]: No.  Deleting.

File: annotation-file-utilities/src/annotator/find/TreeFinder.java (hunk 32 above)
Comment (dbrosoft) [inline annotation-file-utilities/src/annotator/find/TreeFinder.java]: I've added a return field to the javadoc comment.

File: annotation-file-utilities/src/annotator/find/Insertion.java (hunk 19 above)
Comment (dbrosoft) [inline annotation-file-utilities/src/annotator/find/Insertion.java]: Most often, only annotations get inserted.  (CloseParenthesesInsertion is the most obvious exception -- it inserts neither annotations nor types.)

File: annotation-file-utilities/src/annotator/find/Insertion.java (hunk 20 above)
Comment (dbrosoft) [inline annotation-file-utilities/src/annotator/find/Insertion.java]: I meant "generated" in the sense of source code generation.  I suppose "inserted" would be clearer here.

File: annotation-file-utilities/src/annotator/find/InMethodCriterion.java (hunk 3 above)
Comment (dbrosoft) [inline annotation-file-utilities/src/annotator/find/InMethodCriterion.java]: Refactored away in master.

**Comment by mernst** [comment]
No response to extensive review feedback, and original pull request said this feature is likely to be incorrect anyway.  Therefore, I am closing this pull request.
